feat: integrate gray logger to improve scicat error logging

Junjiequan opened the pull request:
## Description

@user-office-software/duo-logger package is installed and gray logger is configured.

## Motivation

Enabling grayLogger allows SIMS team to get error messages in the first place.



## Tests included/Docs Updated?

- [x] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked) 
- [ ] Docs updated?
- [ ] New packages used/requires npm install? 
- [ ] Toggle added for new features?

Commit:
Fixed props of and added HistoryClass to OpenAPI Schema #941

Junjiequan committed:
feat: added grayLogger

Junjiequan committed:
added unit test for logger service

Junjiequan committed:
use inspect to avoid cicular reference error

Junjiequan committed:
added grayLog service env variable

Junjiequan committed:
use gray logger to log ES initial connection

Junjiequan committed:
add gray logger to log Origdatablocks file download

Junjiequan committed:
changed wrong test case name

bpedersen2 commented on src/elastic-search/elastic-search.service.ts:
```diff
@@ -85,9 +90,18 @@ export class ElasticSearchService implements OnModuleInit {
+      logger.logInfo("ElasticSearch Connected.", {
+        host: this.host,
+        username: this.username,
+        password: this.password,
```
Please never log passwords!

martin-trajanovski commented on src/elastic-search/elastic-search.service.ts:
```diff
@@ -69,6 +70,10 @@ export class ElasticSearchService implements OnModuleInit {
       Logger.error(
         "Missing ENVIRONMENT variables for elastic search connection",
       );
+      logger.logError(
```
I am a bit confused here about logging the same message twice with two different loggers. Is this really needed?

bpedersen2 replied:
https://docs.nestjs.com/techniques/logger#using-the-logger-for-application-logging 

gives a way on how to use a single logger instance, soit seems you would never need to directly instantiate a custom class here.

Junjiequan replied:
The idea was to use nestJS logger for development and gray logger to log critical errors on graylogs. I'm not sure if we want to send every errors to grayLog

bpedersen2 replied:
Then maybe check https://docs.nestjs.com/techniques/logger#use-external-logger as well, it recommends to use e.g. winston for production setups ( and winston allows to configure different/multiple loggers and filtering by level etc.)

Junjiequan replied:
winston looks like a good option for configuring external loggers. I will talk about this after Max comes back from the holiday

bpedersen2 commented on src/logger/logger.service.ts:
```diff
@@ -0,0 +1,42 @@
+            this.server,
+            parseInt(this.port),
+            {
+              facility: "DMSC",
```
harcoded facility... needs to be configurable

bpedersen2 committed:
Don't fail if es is not enabled and config is missing

If es is not enabled, startup fails as values for the related vars
are still required.

Change-Id: I23cf5ce862bb503eb308bb7e6431fef51edda291

Junjiequan committed:
make grayLogger facility variable configurable

Junjiequan committed:
removed credentials from logger message

dependabot[bot] committed:
build(deps): bump @nestjs-modules/mailer from 1.9.1 to 1.10.3

Bumps [@nestjs-modules/mailer](https://github.com/nest-modules/mailer) from 1.9.1 to 1.10.3.
- [Release notes](https://github.com/nest-modules/mailer/releases)
- [Changelog](https://github.com/nest-modules/mailer/blob/main/CHANGELOG.md)
- [Commits](https://github.com/nest-modules/mailer/compare/v1.9.1...v1.10.3)

---
updated-dependencies:
- dependency-name: "@nestjs-modules/mailer"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump mathjs from 12.2.0 to 12.2.1

Bumps [mathjs](https://github.com/josdejong/mathjs) from 12.2.0 to 12.2.1.
- [Changelog](https://github.com/josdejong/mathjs/blob/develop/HISTORY.md)
- [Commits](https://github.com/josdejong/mathjs/compare/v12.2.0...v12.2.1)

---
updated-dependencies:
- dependency-name: mathjs
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump openid-client from 5.6.1 to 5.6.2

Bumps [openid-client](https://github.com/panva/node-openid-client) from 5.6.1 to 5.6.2.
- [Release notes](https://github.com/panva/node-openid-client/releases)
- [Changelog](https://github.com/panva/node-openid-client/blob/main/CHANGELOG.md)
- [Commits](https://github.com/panva/node-openid-client/compare/v5.6.1...v5.6.2)

---
updated-dependencies:
- dependency-name: openid-client
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/parser from 6.14.0 to 6.15.0

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 6.14.0 to 6.15.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.15.0/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump eslint-plugin-prettier from 5.0.1 to 5.1.2

Bumps [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier) from 5.0.1 to 5.1.2.
- [Release notes](https://github.com/prettier/eslint-plugin-prettier/releases)
- [Changelog](https://github.com/prettier/eslint-plugin-prettier/blob/master/CHANGELOG.md)
- [Commits](https://github.com/prettier/eslint-plugin-prettier/compare/v5.0.1...v5.1.2)

---
updated-dependencies:
- dependency-name: eslint-plugin-prettier
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 6.14.0 to 6.15.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.15.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/supertest from 2.0.16 to 6.0.1

Bumps [@types/supertest](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/supertest) from 2.0.16 to 6.0.1.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/supertest)

---
updated-dependencies:
- dependency-name: "@types/supertest"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 6.15.0 to 6.16.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.16.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/supertest from 6.0.1 to 6.0.2

Bumps [@types/supertest](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/supertest) from 6.0.1 to 6.0.2.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/supertest)

---
updated-dependencies:
- dependency-name: "@types/supertest"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/node from 20.10.5 to 20.10.6

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 20.10.5 to 20.10.6.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump nodemailer from 6.9.7 to 6.9.8

Bumps [nodemailer](https://github.com/nodemailer/nodemailer) from 6.9.7 to 6.9.8.
- [Release notes](https://github.com/nodemailer/nodemailer/releases)
- [Changelog](https://github.com/nodemailer/nodemailer/blob/master/CHANGELOG.md)
- [Commits](https://github.com/nodemailer/nodemailer/compare/v6.9.7...v6.9.8)

---
updated-dependencies:
- dependency-name: nodemailer
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/node-fetch from 2.6.9 to 2.6.10

Bumps [@types/node-fetch](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node-fetch) from 2.6.9 to 2.6.10.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node-fetch)

---
updated-dependencies:
- dependency-name: "@types/node-fetch"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/parser from 6.15.0 to 6.16.0

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 6.15.0 to 6.16.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.16.0/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

jkotan committed:
workaround for wrong degC to K conversion

nitrosx reviewed (commented):
I feel that this is an ESS specific change, aka graylog.
I think we should have a wider discussion and see if we can generalize it.
Also, I agree with @bpedersen2 that we should have a single logger in SciCat which can be configured to send the log to multiple logging facilities according to the IT infrastructure where the instance is running

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 6.16.0 to 6.18.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.18.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/luxon from 3.3.7 to 3.4.0

Bumps [@types/luxon](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/luxon) from 3.3.7 to 3.4.0.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/luxon)

---
updated-dependencies:
- dependency-name: "@types/luxon"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump openid-client from 5.6.2 to 5.6.4

Bumps [openid-client](https://github.com/panva/node-openid-client) from 5.6.2 to 5.6.4.
- [Release notes](https://github.com/panva/node-openid-client/releases)
- [Changelog](https://github.com/panva/node-openid-client/blob/main/CHANGELOG.md)
- [Commits](https://github.com/panva/node-openid-client/compare/v5.6.2...v5.6.4)

---
updated-dependencies:
- dependency-name: openid-client
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/parser from 6.16.0 to 6.18.0

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 6.16.0 to 6.18.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.18.0/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @nestjs/schematics from 10.0.3 to 10.1.0

Bumps [@nestjs/schematics](https://github.com/nestjs/schematics) from 10.0.3 to 10.1.0.
- [Release notes](https://github.com/nestjs/schematics/releases)
- [Changelog](https://github.com/nestjs/schematics/blob/master/.release-it.json)
- [Commits](https://github.com/nestjs/schematics/compare/10.0.3...10.1.0)

---
updated-dependencies:
- dependency-name: "@nestjs/schematics"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/node from 20.10.6 to 20.10.7

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 20.10.6 to 20.10.7.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @nestjs/cli from 10.2.1 to 10.3.0

Bumps [@nestjs/cli](https://github.com/nestjs/nest-cli) from 10.2.1 to 10.3.0.
- [Release notes](https://github.com/nestjs/nest-cli/releases)
- [Changelog](https://github.com/nestjs/nest-cli/blob/master/.release-it.json)
- [Commits](https://github.com/nestjs/nest-cli/compare/10.2.1...10.3.0)

---
updated-dependencies:
- dependency-name: "@nestjs/cli"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/passport-jwt from 3.0.13 to 4.0.0

Bumps [@types/passport-jwt](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/passport-jwt) from 3.0.13 to 4.0.0.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/passport-jwt)

---
updated-dependencies:
- dependency-name: "@types/passport-jwt"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Junjiequan committed:
improved error logs integration with grayLogger

dependabot[bot] committed:
build(deps): bump follow-redirects from 1.15.2 to 1.15.4

Bumps [follow-redirects](https://github.com/follow-redirects/follow-redirects) from 1.15.2 to 1.15.4.
- [Release notes](https://github.com/follow-redirects/follow-redirects/releases)
- [Commits](https://github.com/follow-redirects/follow-redirects/compare/v1.15.2...v1.15.4)

---
updated-dependencies:
- dependency-name: follow-redirects
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Junjiequan committed:
fixed unit test failing

Junjiequan committed:
renaming customized logger clsas

Junjiequan committed:
minor elasticsearch log message change

Junjiequan committed:
added flag for gray logger levels

nitrosx committed:
added endpoint and related tests

nitrosx committed:
addedd output dtos and updated swagger interface for users

nitrosx committed:
fixed linting issues

nitrosx committed:
fixed linting issue

Junjiequan committed:
created scicatLogger class to instantiate multiple loggers

Junjiequan committed:
defined types for logger methods, cleaned up unnecessary env variables

nitrosx committed:
Merge branch 'master' into SWAP-3353-scicat-be-send-datasets-access-and-download-to-gr

bpedersen2 reviewed (commented):
Question: does it fallback to use the default logger with out a loggers.json present,  or is minimal loggers.json required?

nitrosx commented on src/loggers/logger.service.ts:
```diff
@@ -0,0 +1,127 @@
+    if (fs.existsSync(filePath)) {
+      const data = fs.readFileSync(filePath, "utf8");
+      loggerConfigs = JSON.parse(data);
+    }
```
I would move this section of code in the configuration controller. I would like to follow the concept that if you instantiate the configuration class, you have the full configuration and avoid loading portions of the configuration in other classes.
@bpedersen2 thoughts?

bpedersen2 replied:
Yes, sounds better. Then we can ensure useful defaults.

nitrosx commented on src/elastic-search/elastic-search.service.ts:
```diff
@@ -161,10 +162,13 @@ export class ElasticSearchService implements OnModuleInit {
       return HttpStatus.CREATED;
     } catch (error) {
-      Logger.error("createIndex failed-> ElasticSearchService", error);
+      console.error("createIndex failed-> ElasticSearchService", error);
```
Why not using Logger here too?

Junjiequan replied:
This is for our use case specific change, if you see the catch block it contains log errors and throw errors. Since

nitrosx commented on src/elastic-search/elastic-search.service.ts:
I was wondering why you are using `console.log` and not Logger?

Junjiequan replied:
if we set both logger and `throw` here, two identical errors will be sent to graylogger, one from logError and one from exception. 
Previously I've included console because `throw` was not showing any logs. It is now fixed, and Loggers from all catch blocks are removed.

nitrosx commented on src/loggers/logger.service.ts:
```diff
@@ -0,0 +1,127 @@
+    }
+  }
+
+  private handleLogForwarding(method: string, ...args: unknown[]): void {
```
I was wondering why you are doing this dynamic mapping of the methods instead of declaring an abstract class and make sure that all the loggers implement such class?

nitrosx commented on src/config/configuration.ts:
```diff
@@ -37,6 +35,7 @@ const configuration = () => {
   // Logger.log("- Update job groups : " + updateJobGroups);
 
   return {
+    loggerConfigs: [],
```
Why don't we load the logger configuration in this class instead doing it in the logger class below?
Please see my comment below on logger.service.ts

Junjiequan commented:
> Question: does it fallback to use the default logger with out a loggers.json present, or is minimal loggers.json required?

@bpedersen2  minimal loggers.json is required to enable default logger. and the configuration can be found in loggers.json.example

Junjiequan committed:
changes based on comments

Junjiequan committed:
fix logger service unit test

nitrosx commented on src/config/configuration.ts:
```diff
@@ -22,10 +22,26 @@ const configuration = () => {
+      jsonConfigMap[fileName] = JSON.parse(data);
+      break;
+    }
+  }
```
I would make it more generic, so we can easily use it for the functional accounts and future development.
How about something like this:
```typescript
  const jsonConfigFileList = {
    "loggers" : process.env.LOGGERS_CONFIG_FILE || "loggers.json",
  };
 Object.keys(jsonConfigFileList).forEach( key => {
    const filePath = jsonConfigFileList[key];
    if (fs.existsSync(filePath)) {
      const data = fs.readFileSync(filePath, "utf8");
      jsonConfigMap[fileName] = JSON.parse(data);
      break;
    }
  });
```

nitrosx committed:
Merge branch 'master' into SWAP-3353-scicat-be-send-datasets-access-and-download-to-gr

Junjiequan committed:
change jsonconfig to be more generic

bpedersen2 commented on src/config/configuration.ts:
```diff
@@ -53,7 +61,7 @@ const configuration = () => {
 
   return {
-    loggerConfigs: jsonConfigMap.loggers ?? [defaulgLogger],
+    loggerConfigs: jsonConfigMap.loggers || [defaulgLogger],
```
s/defaulgLogger/defaultLogger/ (and find other instances of that typo)?

Junjiequan committed:
typo fix

Junjiequan committed:
Merge branch 'master' into SWAP-3353-scicat-be-send-datasets-access-and-download-to-gr

Junjiequan committed:
Merge branch 'master' into SWAP-3353-scicat-be-send-datasets-access-and-download-to-gr

Junjiequan committed:
Merge branch 'master' into SWAP-3353-scicat-be-send-datasets-access-and-download-to-gr
